hbase-coprocessor/src/main/java/org/apache/omid/transaction/TransactionVisibilityFilter.java

```diff
@@ -0,0 +1,248 @@
+    }
+
+    public TransactionVisibilityFilter(@Nullable Filter cellFilter,
+                                       SnapshotFilterImpl snapshotFilter,
+                                       HBaseTransaction hbaseTransaction) {
+        this.userFilter = cellFilter;
+        this.snapshotFilter = snapshotFilter;
+        shadowCellCache = new HashMap<>();
+        this.hbaseTransaction = hbaseTransaction;
+        familyDeletionCache = new HashMap<String, Long>();
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell v) throws IOException {
+        if (CellUtils.isShadowCell(v)) {
+            Long commitTs =  Bytes.toLong(CellUtil.cloneValue(v));
+            shadowCellCache.put(v.getTimestamp(), commitTs);
+            // Continue getting shadow cells until one of them fits this transaction
+            if (hbaseTransaction.getStartTimestamp() >= commitTs) {
+                return ReturnCode.NEXT_COL;
+            } else {
+                return ReturnCode.SKIP;
+            }
+        } else if (CellUtils.isFamilyDeleteCell(v)) {
+            //Delete is part of this transaction
+            if (snapshotFilter.getTSIfInTransaction(v, hbaseTransaction).isPresent()) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), v.getTimestamp());
+                return ReturnCode.NEXT_COL;
+            }
+
+            if (shadowCellCache.containsKey(v.getTimestamp()) &&
+                    hbaseTransaction.getStartTimestamp() >= shadowCellCache.get(v.getTimestamp())) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), shadowCellCache.get(v.getTimestamp()));
+                return ReturnCode.NEXT_COL;
+            }
+
+            // Try to get shadow cell from region
+            final Get get = new Get(CellUtil.cloneRow(v));
+            get.setTimeStamp(v.getTimestamp()).setMaxVersions(1);
+            get.addColumn(CellUtil.cloneFamily(v), CellUtils.addShadowCellSuffix(CellUtils.FAMILY_DELETE_QUALIFIER));
+            Result deleteFamilySC = snapshotFilter.getTableAccessWrapper().get(get);
+
+            if (!deleteFamilySC.isEmpty() &&
+                    Bytes.toLong(CellUtil.cloneValue(deleteFamilySC.rawCells()[0] )) < hbaseTransaction.getStartTimestamp()){
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), Bytes.toLong(CellUtil.cloneValue(deleteFamilySC.rawCells()[0])));
+                return ReturnCode.NEXT_COL;
+            }
+
+            //At last go to commit table
+            Optional<Long> commitTimestamp = snapshotFilter.tryToLocateCellCommitTimestamp(hbaseTransaction.getEpoch(),
+                    v, shadowCellCache);
+            if (commitTimestamp.isPresent() && hbaseTransaction.getStartTimestamp() >= commitTimestamp.get()) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), commitTimestamp.get());
+                return ReturnCode.NEXT_COL;
+            }
+
```

Comment:
We should recheck shadow cell in here. For cases that someone updated the shadow cell and remove the entry from the commit table between lines 87 and 94.